Full TS and WebSockets

aimbjp opened the pull request:
Redux store has been typed.

Screens "Order Feed" and "Order History", as well as an order page, have been created.

Routing has been adjusted for these pages.

WebSocket implementation has been added to fetch order data from the server.

Redux store has been implemented to manage the WebSocket state.

aimbjp committed:
Redux store has been typed.

Screens "Order Feed" and "Order History", as well as an order page, have been created.

Routing has been adjusted for these pages.

WebSocket implementation has been added to fetch order data from the server.

Redux store has been implemented to manage the WebSocket state.

gennady-bars commented on src/services/hooks.ts:
```diff
@@ -0,0 +1,10 @@
+import { AppDispatch, RootState } from './types';
+
+
+export const useDispatch: () => AppDispatch = dispatchHook;
+export const useSelector: TypedUseSelectorHook<RootState> = selectorHook;
```
* Отлично типизировали хуки `useSelector` и `useDispatch`
* Отлично типизировали стейт, диспатч

gennady-bars commented on src/services/middleware/socket-middleware.ts:
```diff
@@ -0,0 +1,69 @@
+import {TWSActions} from "../types/websocket";
+import {BURGER_ORDERS_WS_URL, refreshToken} from "../api/api-norma";
+
+export const socketMiddleware = (wsActions: TWSStoreActions, withTokenRefresh: boolean = false): Middleware => {
+    return ((store: MiddlewareAPI<AppDispatch, RootState>) => {
```
Хорошо сделан `socketMiddleware`

gennady-bars commented on src/components/profile/profile-feed.tsx:
```diff
@@ -0,0 +1,25 @@
+            dispatch({type: WS_CONNECTION_START, payload: (`${BURGER_ORDERS_WS_URL}?token=${localStorage.getItem('accessToken')?.split(' ')[1]}`)
+            });
+            return () => {
+                dispatch({type: WS_CONNECTION_CLOSE});
```
Отлично, что закрываете соединение сокетов при покидании страниц

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -1,21 +1,8 @@
+import {useSelector} from "../../services/hooks";
+import styleLoader from '../../utils/styles/loader.module.css';
+import {Loader} from "../elements";
 
```
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -54,6 +58,30 @@ function App() {
+                    />
+                </Routes>
+            )}
+            {background && (
+                <Routes>
+                    <Route
+                        path='/profile/orders/:id'
+                        element={
+                            <Modal title="Заказ" onClose={handleCloseModal}>
+                                <Order />
```
* Если **перезагрузить** страницу с открытым попапом заказа **истории**, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки. 


Для попапа истории нужно использовать `OnlyAuth`, ведь туда должны опадать только авторизованные пользователи

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично типизировали хуки `useSelector` и `useDispatch`
* Отлично типизировали стейт, диспатч
* Хорошо сделан `socketMiddleware`
* Отлично, что закрываете соединение сокетов при покидании страниц 
* Хорошо типизировали редакс


но есть некоторые недочеты:




* В ленте ничего не отображается. Все время загрузка https://disk.yandex.ru/i/5-NDZ1QSFtMcvA . Но после авторизации отобразилось, хотя авторизация не нужна для ленты
* Если **перезагрузить** страницу с открытым попапом заказа **истории**, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки. 
* Для попапа истории нужно использовать `OnlyAuth`, ведь туда должны опадать только авторизованные пользователи. Поэтому и не работает логика с перезагрузкой



**Можно лучше**

* При открытии попапа в консоли много предупреждений https://disk.yandex.ru/i/_9_UuTh2_FruUQ 
* Попапы нужно сделать по макету

aimbjp committed:
Fix routes (make only auth users getting info about order from profile, cause we need access token for this)

Fix connection to websocket, so everyone can connect to it now and feed is appearing successfully

aimbjp commented:
`При открытии попапа в консоли много предупреждений https://disk.yandex.ru/i/_9_UuTh2_FruUQ
`
Не получилось найти информацию по этому поводу, тоже попробую позже.
Мб у Вас есть какие-то предположения или наводки, из-за чего это может происходить?

gennady-bars commented:
> Мб у Вас есть какие-то предположения или наводки, из-за чего это может происходить?

 Обычно это решается тем, что добавляется в конце `<Route path='*' ... />`, чтобы любой рут обрабатывался, но тут не срабатывает это...

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.



* Попап с номером заказа нужно сделать по макету https://disk.yandex.ru/i/LtrYCV-OkOSyZw
